### PR TITLE
Improve tracing and add tracing to Wan

### PR DESCRIPTION
**Pull request by friedrich**
### Ticket

https://github.com/tenstorrent/tt-metal/issues/36032

### What's new

This PR add support for tracing to the whole Wan pipeline (encoder, DiT, VAE). Moreover it adds several tracing improvements to tt_dit:

* Conv2d tracing support.
* Replacements for `ttnn.full`, `ttnn.arange`, `ttnn.tril`, and `ttnn.triu` that are compatible with tracing.
* The function `utils.tensor.pad_single` which transparently (and I think without performance loss) works around the fact that `ttnn.pad` does not allow padding of arbitrary tensor dimension.
* The `Tracer` class now has safer defaults that can all be overridden: clones input tensors for the compile run, executes trace after compile run.

### Checklist

- [ ] [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:friedrich/tracing-wan)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:friedrich/tracing-wan)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:friedrich/tracing-wan)
- [x] New/Existing tests provide coverage for changes
- [x] (Optional) Ran [clang-tidy code analysis](https://github.com/tenstorrent/tt-metal/actions/workflows/code-analysis.yaml) on the PR branch to catch linting errors early

#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:friedrich/tracing-wan)
  - [ ] `models-mandatory` preset (runs: [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [x] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:friedrich/tracing-wan)
  - [ ] `models-mandatory` preset (runs: [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [x] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:friedrich/tracing-wan)
  - [ ] `models-mandatory` preset (runs: [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [x] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

### CODE REVIEW

File: models/tt_dit/pipelines/wan/pipeline_wan.py [hunk 1]
```diff
@@ -267,18 +270,32 @@ def __init__(
...
+
+    def _allocate_persistent_buffers(self) -> None:
+        """Allocate persistent buffers by running a pipeline pass without tracing.
+
+        This is important so they do not get allocated after trace capture, which would lead to
+        them being overwritten during trace execution. The resolution must match the resolution
+        used for traced inference.
+        """
+        logger.info("Pipeline allocation run...")
+        self.run_single_prompt(prompt="", num_inference_steps=2, seed=0, traced=False, num_frames=1)
```
Comment (cglagovichTT) [inline models/tt_dit/pipelines/wan/pipeline_wan.py]: This call in allocate must have the same output shape (num_frames, height, width) as the pipeline will be called with later, right?

Comment (friedrich) [reply models/tt_dit/pipelines/wan/pipeline_wan.py]: In principle yes, but the way the model is currently implemented, the allocation run is not strictly needed, as every traced part is retraced after another one runs. I will check if we can just remove it.

File: models/tt_dit/pipelines/wan/pipeline_wan.py [hunk 2]
```diff
@@ -1025,3 +1101,79 @@ def run_single_prompt(self, *args, **kwargs):
...
+            rope_cos_1HND,
+            rope_sin_1HND,
+            trans_mat,
+            N,
+            temb_11BD,
+            timestep_proj_1BTD,
+        )
+        return cond, uncond
+
+    def _transformer_2_step(
```
Comment (sadesoyeTT) [inline models/tt_dit/pipelines/wan/pipeline_wan.py]: Can we collapse this into _transformer_step. And just pass the corresponding transformer as a parameter?

File: models/tt_dit/models/vae/vae_wan2_1.py [hunk 3]
```diff
@@ -1118,7 +1081,8 @@ def __init__(
             out_dim,
             out_channels,
             kernel_size=3,
-            padding=1,
+            padding=(2, 1, 1),
```
Comment (sadesoyeTT) [inline models/tt_dit/models/vae/vae_wan2_1.py:1087]: Any reason for the padding changes?

File: models/tt_dit/pipelines/wan/pipeline_wan.py [hunk 4]
```diff
@@ -1014,6 +1086,10 @@ def __call__(
 
         # Offload all models
         # self.maybe_free_model_hooks()
+        self.transformer.cached_rope_features.clear()
```
Comment (sadesoyeTT) [inline models/tt_dit/pipelines/wan/pipeline_wan.py]: Any issues with reusing for subsequent calls?

File: models/tt_dit/tests/models/wan2_2/test_pipeline_wan.py [hunk 5]
```diff
@@ -12,16 +12,18 @@
...
 
+DEVICE_PARAMS = {"trace_region_size": 90000000}
+
 
 @pytest.mark.parametrize(
     "mesh_device, mesh_shape, sp_axis, tp_axis, num_links, dynamic_load, device_params, topology, is_fsdp",
     [
-        [(2, 2), (2, 2), 0, 1, 2, False, line_params, ttnn.Topology.Linear, True],
-        [(2, 4), (2, 4), 0, 1, 1, True, line_params, ttnn.Topology.Linear, True],
+        ((2, 2), (2, 2), 0, 1, 2, False, {**DEVICE_PARAMS, **line_params}, ttnn.Topology.Linear, True),
```
Comment (sadesoyeTT) [inline models/tt_dit/tests/models/wan2_2/test_pipeline_wan.py:21]: Please move contents of DEVICE_PARAMS into the line and ring_params to clean this up some.

File: models/tt_dit/tests/models/wan2_2/test_pipeline_wan.py [hunk 6]
```diff
@@ -42,7 +44,15 @@
         "resolution_720p",
     ],
 )
+@pytest.mark.parametrize(
+    "traced",
+    [
+        pytest.param(True, id="tracing_on"),
```
Comment (sadesoyeTT) [inline models/tt_dit/tests/models/wan2_2/test_pipeline_wan.py:50]: Can we change to traced and not_traced to align with other models.

File: models/tt_dit/tests/models/wan2_2/test_transformer_wan.py [hunk 7]
```diff
@@ -380,14 +380,15 @@ def test_wan_transformer_inner_step(
 
     # Run TT inner_step (returns on-device tensor)
     logger.info(f"Running TT inner_step with spatial_host shape {spatial_host.shape}, N={N}")
+    spatial_1BNI = bf16_tensor(
```
Comment (sadesoyeTT) [inline models/tt_dit/tests/models/wan2_2/test_transformer_wan.py:383]: Thanks!

File: models/tt_dit/tests/unit/test_tensor_utils.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
+# SPDX-FileCopyrightText: © 2026 Tenstorrent AI ULC
```
Comment (sadesoyeTT) [inline models/tt_dit/tests/unit/test_tensor_utils.py:1]: If this is specifically to test tracing on tensors, we should add comment or update file/var names as needed to indicate such.

File: models/tt_dit/utils/tensor.py [hunk 9]
```diff
@@ -268,3 +307,104 @@ def unflatten(x: ttnn.Tensor, dim: int, sizes: Sequence[int]) -> ttnn.Tensor:
     else:
         new_shape[dim : dim + 1] = sizes
     return ttnn.reshape(x, new_shape)
+
+
+def full(
```
Comment (sadesoyeTT) [inline models/tt_dit/utils/tensor.py:314]: What's the difference with this and ttnn.full(...,fill_value, layour, device,...)?

File: models/tt_dit/utils/tensor.py [hunk 10]
```diff
@@ -268,3 +307,104 @@ def unflatten(x: ttnn.Tensor, dim: int, sizes: Sequence[int]) -> ttnn.Tensor:
...
+    """Alternative to `ttnn.full` that supports tracing."""
+    if not isinstance(size, ttnn.Shape):
+        size = ttnn.Shape(size)
+
+    result = ttnn.allocate_tensor_on_device(size, dtype, layout, device, memory_config)
+    ttnn.fill(result, fill_value, output_tensor=result)
+    return result
+
+
+def arange(
```
Comment (sadesoyeTT) [inline models/tt_dit/utils/tensor.py:332]: Similar to full. Why not use https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/api/ttnn.arange.html#ttnn.arange ?

File: models/tt_dit/pipelines/wan/pipeline_wan.py [hunk 11]
```diff
@@ -386,8 +406,15 @@ def _prepare_text_encoder(self):
             mesh_shape=tuple(self.mesh_device.shape),
             get_torch_state_dict=lambda: self.text_encoder.state_dict(),
         )
+        if self._encoder_tracer is None:
+            self._encoder_tracer = Tracer(
+                self.tt_umt5_encoder.forward, device=self.mesh_device, clone_prep_inputs=False
+            )
 
     def _prepare_transformer1(self):
+        if self._transformer_tracer is not None:
```
Comment (sadesoyeTT) [inline models/tt_dit/pipelines/wan/pipeline_wan.py]: This would incur tracing overhead every time potentially making things slower. It may be best to just disable tracing for dynamic loading.

File: models/tt_dit/pipelines/wan/pipeline_wan.py [hunk 12]
```diff
@@ -418,12 +457,15 @@ def _prepare_vae(self):
             mesh_shape=tuple(self.mesh_device.shape),
             get_torch_state_dict=lambda: self.vae.state_dict(),
         )
+        if self._vae_tracer is None:
+            self._vae_tracer = Tracer(self.tt_vae.forward, device=self.mesh_device, clone_prep_inputs=False)
```
Comment (sadesoyeTT) [inline models/tt_dit/pipelines/wan/pipeline_wan.py:455]: How easy would it be to plug the tracer into the Module class instead? We can plan it for another time.

File: models/tt_dit/pipelines/wan/pipeline_wan.py (hunk 11 above)
Comment (sadesoyeTT) [reply models/tt_dit/pipelines/wan/pipeline_wan.py]: For VAE and text encoder. Transformer tracing is correct.

File: models/tt_dit/models/vae/vae_wan2_1.py (hunk 3 above)
Comment (friedrich) [reply models/tt_dit/models/vae/vae_wan2_1.py:1087]: Yes, diffusers lies about padding. Since it is front padding only, they multiply by two, which makes it impossible to have odd padding. This is explained in the docstring of `WanCausalConv3d.__init__`.

File: models/tt_dit/pipelines/wan/pipeline_wan.py (hunk 11 above)
Comment (friedrich) [reply models/tt_dit/pipelines/wan/pipeline_wan.py]: Totally agree. I just let it enabled to verify it works on T3K. I was also thinking if we could reuse the trace even when model weights are unloaded. Maybe by reserving a chunk of memory early on, so that weights get a deterministic location in memory.

File: models/tt_dit/pipelines/wan/pipeline_wan.py (hunk 12 above)
Comment (friedrich) [reply models/tt_dit/pipelines/wan/pipeline_wan.py:455]: Probably easy. One issue though is that we often do not trace *just* the model, so I think it is good to preserve the flexibility of a separate tracer class. Maybe the way to go is to make it a mixin that models use that are intended to be traced directly. This way we would also keep `Module` simple.

File: models/tt_dit/tests/models/wan2_2/test_pipeline_wan.py (hunk 6 above)
Comment (friedrich) [reply models/tt_dit/tests/models/wan2_2/test_pipeline_wan.py:50]: Yeah we have a bunch of different conventions. We also have "yes_traced" and "no_traced". I agree this should be unified but I think it is convenient to have something else than just "traced" because when pytest is called with "-k traced" it matches both, the traced and the untraced version. What I do is call it with "-k -traced" which matches only the traced version but looks very weird. What do you think?

File: models/tt_dit/tests/unit/test_tensor_utils.py (hunk 8 above)
Comment (friedrich) [reply models/tt_dit/tests/unit/test_tensor_utils.py:1]: Not quite. It also test `test_pad_single` which is not specific to tracing, it is just tested with tracing on to make sure it also works with it. So the idea is to test anything in `utils/tensor.py`, it just happens that most of the things currently tested are related to tracing.

File: models/tt_dit/utils/tensor.py (hunk 9 above)
Comment (friedrich) [reply models/tt_dit/utils/tensor.py:314]: `ttnn.full` can not be used with tracing in my experience.

File: models/tt_dit/tests/models/wan2_2/test_pipeline_wan.py (hunk 5 above)
Comment (friedrich) [reply models/tt_dit/tests/models/wan2_2/test_pipeline_wan.py:21]: Are you sure we want them in there? They are used all over the place.

File: models/tt_dit/pipelines/wan/pipeline_wan.py (hunk 4 above)
Comment (friedrich) [reply models/tt_dit/pipelines/wan/pipeline_wan.py]: Cleared it for free up memory, but seems fine without it. Removed.

File: models/tt_dit/pipelines/wan/pipeline_wan.py (hunk 1 above)
Comment (friedrich) [reply models/tt_dit/pipelines/wan/pipeline_wan.py]: I removed it. Currently it is not useful, since we need to release persistent CCL buffers between trace captures to free up space, which makes the CCL manager call `synchronize_device` again. So we need to do an allocation run before any trace capture anyway.

File: models/tt_dit/pipelines/wan/pipeline_wan.py (hunk 2 above)
Comment (friedrich) [reply models/tt_dit/pipelines/wan/pipeline_wan.py]: Done

File: models/tt_dit/utils/tensor.py (hunk 10 above)
Comment (friedrich) [reply models/tt_dit/utils/tensor.py:332]: Same reason as with `ttnn.full`.

File: models/tt_dit/pipelines/wan/pipeline_wan.py (hunk 11 above)
Comment (friedrich) [reply models/tt_dit/pipelines/wan/pipeline_wan.py]: I now disabled DiT tracing if dynamic loading is on in the pipeline.

File: models/tt_dit/tests/models/wan2_2/test_pipeline_wan.py (hunk 5 above)
Comment (sadesoyeTT) [reply models/tt_dit/tests/models/wan2_2/test_pipeline_wan.py:21]: Yea, you are right. I'm used to seeing them only in Wan, so bad assumption on my end. We can have specific params for wan. But not top priority.

File: models/tt_dit/tests/models/wan2_2/test_pipeline_wan.py (hunk 6 above)
Comment (sadesoyeTT) [reply models/tt_dit/tests/models/wan2_2/test_pipeline_wan.py:50]: Yea, that make sense. I always use weird -k "not traced". But we can pick a convention that's easier to deal with and adapt all. Perhaps we can do that when its time to update CI test.

File: models/tt_dit/utils/tensor.py (hunk 9 above)
Comment (sadesoyeTT) [reply models/tt_dit/utils/tensor.py:314]: Interesting. We should keep an eye for it and possibly file ttnn issue. I vaguely recall we use full for the timestep and sigma in sd35 tracing, but perhaps that's a different code path.

File: models/tt_dit/pipelines/wan/pipeline_wan.py (hunk 12 above)
Comment (sadesoyeTT) [reply models/tt_dit/pipelines/wan/pipeline_wan.py:455]: Agreed. The reason for the suggestion was because of the necessary management stuff that needs to be done upon loading and deallocating models. Having a central functionality that takes care of that sounds good to me.

File: models/tt_dit/utils/tensor.py (hunk 9 above)
Comment (friedrich) [reply models/tt_dit/utils/tensor.py:314]: Right, we have `ttnn.full` in the SD3.5 loop but its outside the traced region.

**Comment by friedrich** [comment]
Superseded by https://github.com/tenstorrent/tt-metal/pull/41384 and https://github.com/tenstorrent/tt-metal/pull/40624 and
